parsers/pascal.c

```diff
@@ -74,6 +80,68 @@ static bool tail (const char *cp)
 	return result;
 }
 
+static void parseArglist(const char *buf, char **arglist, char **vartype)
+{
+	char *c, *start, *end;
+	int level;
+
+	if (NULL == buf || NULL == arglist)
+		return;
+
+	c = strdup(buf);
+	/* parse argument list which can be missing like in "function ginit:integer;" */
+	if (NULL != (start = strchr(c, '(')))
+	{
+		for (level = 1, end = start + 1; level > 0; ++end)
+		{
+			if ('\0' == *end)
+				break;
+			else if ('(' == *end)
+				++ level;
+			else if (')' == *end)
+				-- level;
+		}
+	}
+	else /* if no argument list was found, continue looking for a return value */
+	{
+		start = "()";
+		end = c;
+	}
+
+	/* parse return type if requested by passing a non-NULL vartype argument */
+	if (NULL != vartype)
+	{
+		char *var, *var_start;
+
+		*vartype = NULL;
+
+		if (NULL != (var = strchr(end, ':')))
+		{
+			var++; /* skip ':' */
+			while (isspace((int) *var))
+				++var;
+
+			if (starttoken(*var))
+			{
+				var_start = var;
+				var++;
+				while (intoken(*var))
+					var++;
+				if (endtoken(*var))
+				{
+					*var = '\0';
+					*vartype = strdup(var_start);
+				}
+			}
+		}
+	}
+
+	*end = '\0';
+	*arglist = strdup(start);
+
+	eFree(c);
```

Comment:
If `eStrndup` helps you expectedly, you can remove this `eFree()`.

Reply:
Fortunately, that code vanished by using `vString`.